host/rpc.go

```diff
@@ -0,0 +1,213 @@
+
+	budget, err := sh.processPayment(stream)
+	if err != nil {
+		sh.log.Warnln("RPC account balance:", "failed to process payment:", err)
+		rpc.WriteResponse(stream, nil, err)
+		return
+	}
+	defer budget.Refund()
+
+	if err := budget.Spend(settings.RPCFundAccountCost); err != nil {
+		sh.log.Warnln("RPC account balance:", "failed to pay for fund account RPC:", err)
+		rpc.WriteResponse(stream, nil, err)
+		return
+	}
+
+	var req rhp.RPCFundAccountRequest
+	if err = rpc.ReadResponse(stream, &req); err != nil {
+		sh.log.Warnln("RPC account balance:", "failed to read fund account request:", err)
+		rpc.WriteResponse(stream, nil, errors.New("failed to read fund account request"))
+		return
+	}
+
+	fundAmount := budget.Remaining()
+	balance, err := sh.accounts.Deposit(req.AccountID, fundAmount)
+	if err != nil {
+		sh.log.Warnln("RPC account balance:", "failed to fund account:", err)
+		rpc.WriteResponse(stream, nil, errors.New("failed to fund account"))
+		return
+	}
+	if err := budget.Spend(fundAmount); err != nil {
+		sh.log.Warnln("RPC account balance:", "failed to spend account funding:", err)
+		rpc.WriteResponse(stream, nil, err)
+		return
+	}
+
+	resp := &rhp.RPCFundAccountResponse{
+		Balance: balance,
+		Receipt: rhp.Receipt{
+			Host:      sh.public,
+			Account:   req.AccountID,
+			Amount:    fundAmount,
+			Timestamp: time.Now(),
+		},
+	}
+
+	h := types.NewHasher()
+	resp.Receipt.EncodeTo(h.E)
+	resp.Signature = types.SignHash(sh.secret, h.Sum())
+
+	// write the receipt and current balance.
+	err = rpc.WriteResponse(stream, resp, nil)
+	if err != nil {
+		sh.log.Warnln("RPC account balance:", "failed to write fund account response:", err)
+		return
+	}
+}
+
+func (sh *SessionHandler) handleRPCLatestRevision(stream *mux.Stream) {
+	var settingsUID types.UID
+	if err := rpc.ReadObject(stream, &settingsUID); err != nil {
+		sh.log.Warnln("failed to read settings UID:", err)
+		return
+	}
+
+	settings, err := sh.validSettings(settingsUID)
+	if err != nil {
+		sh.log.Warnf("settings uid %s not found:", settingsUID)
+		rpc.WriteResponse(stream, nil, err)
+		return
+	}
+
+	budget, err := sh.processPayment(stream)
+	if err != nil {
+		sh.log.Warnln("failed to process payment:", err)
+		rpc.WriteResponse(stream, nil, err)
+		return
+	}
+	defer budget.Refund()
+
+	if err := budget.Spend(settings.RPCLatestRevisionCost); err != nil {
+		sh.log.Warnln("failed to pay for latest revision RPC:", err)
+		rpc.WriteResponse(stream, nil, err)
+		return
+	}
+}
+
+func (sh *SessionHandler) handleRPCSettings(stream *mux.Stream) {
+	settings := sh.settings.Settings()
+	_, _ = frand.Read(settings.UID[:])
+
+	buf, err := json.Marshal(settings)
+	if err != nil {
+		sh.log.Errorln("failed to marshal settings:", err)
+		return
+	}
+
+	// write the price table to the stream. The price table is sent
+	// before payment so the renter can determine if they want to
+	// continue interacting with the host.
+	err = rpc.WriteResponse(stream, &rhp.RPCSettingsResponse{
```

Comment:
something feels weird here. The host sends its settings to the renter, then waits for the renter to pay it, but if the renter just disconnects without paying, it's fine?

Did the original code send the prices first, then wait for payment, then send the full settings?

Reply:
This follows the siad RHP3 settings RPC, but it only ever returned the price table. I combined the RHP2 external settings and the RHP3 price table into `HostSettings`. 

The renter needs the settings first to know how much to pay for "registering" the settings. The renter can choose not to pay if it decides the prices are too high.

Reply:
Maybe withhold the settings ID until after payment, then? I don't like writing an empty struct to signal success, feels janky.

I'm not in love with having a settings table in general; host prices don't change that frequently, and when they do, the host can return an "outdated settings" error, the renter can re-request the new settings, and then retry. Whereas with locked-in settings, you have to maintain an arbitrarily-sized map in memory with expiration dates.

Reply:
I agree I don't think we necessarily need to register the price tables, and we can get by with short-lived settings.

Reply:
yeah, historically we've had lots of weird bugs where the host would just return "low paying output" or something and it wasn't clear if the duration was wrong, or if the settings changed, or something else. Would definitely like to get to the bottom of that.